text/0000-non-exhaustive.md

```diff
@@ -0,0 +1,548 @@
+
+An attribute `#[non_exhaustive]` is added to the language, which will (for now)
+fail to compile if it's used on anything other than an enum, struct definition,
+or enum variant.
+
+## Enums
+
+Within the crate that defines the enum, this attribute is essentially ignored,
+so that the current crate can continue to exhaustively match the enum. The
+justification for this is that any changes to the enum will likely result in
+more changes to the rest of the crate. Consider this example:
+
+```rust
+use std::error::Error as StdError;
+
+#[non_exhaustive]
+pub enum Error {
+    Message(String),
+    Other,
+}
+impl StdError for Error {
+    fn description(&self) -> &str {
+        match *self {
+            Message(ref s) => s,
+            Other => "other or unknown error",
+        }
+    }
+}
+```
+
+It seems undesirable for the crate author to use a wildcard arm here, to
+ensure that an appropriate description is given for every variant. In fact, if
+they use a wildcard arm in addition to the existing variants, it should be
+identified as dead code, because it will never be run.
+
+Outside the crate that defines the enum, users should be required to add a
+wildcard arm to ensure forward-compatibility, like so:
+
+```rust
+use mycrate::Error;
+
+match error {
+    Message(ref s) => ...,
+    Other => ...,
+    _ => ...,
+}
+```
+
+And it should *not* be marked as dead code, even if the compiler does mark it as
+dead and remove it.
+
+Note that this can *potentially* cause breaking changes if a user adds
+`#[deny(dead_code)]` to a match statement *and* the upstream crate removes the
+`#[non_exhaustive]` lint. That said, modifying warn-only lints is generally
+assumed to not be a breaking change, even though users can make it a breaking
+change by manually denying lints.
+
+## Structs
+
+Like with enums, the attribute is essentially ignored in the crate that defines
+the struct, so that users can continue to construct values for the struct.
+However, this will prevent downstream users from constructing or exhaustively
+matching the struct, because fields may be added to the struct in the future.
+
+Additionally, adding `#[non_exhaustive]` to an enum variant will operate exactly
+the same as if the variant were a struct.
+
+Using our `Config` again:
+
+```rust
+#[non_exhaustive]
+pub struct Config {
+    pub window_width: u16,
+    pub window_height: u16,
+}
+```
+
+We can still construct our config within the defining crate like so:
+
+```rust
+let config = Config { window_width: 640, window_height: 480 };
+```
+
+And we can even exhaustively match on it, like so:
+
+```rust
+if let Ok(Config { window_width, window_height }) = load_config() {
+    // ...
+}
+```
+
+But users outside the crate won't be able to construct their own values, because
+otherwise, adding extra fields would be a breaking change.
+
+Users can still match on `Config`s non-exhaustively, as usual:
+
+```rust
+let &Config { window_width, window_height, .. } = config;
```

Comment:
The RFC doesn't talk about the record update in expressions (FRU) - `let x = Config { window_width, window_height, ..other_config }`, but it probably should.
FRU in expressions, like `..` in patterns, doesn't break the struct's extensibility and may be quite useful.

Reply:
It seems that this thread is why the following doesn't work:

```rust
// Crate 1
#[non_exhaustive]
pub struct Options {
    pub foo: String,
    pub bar: String,
}

impl Default for Options {
    fn default() -> Self {
        Options { foo: "foo".to_string(), bar: "bar".to_string() }
    }
}

// Crate 2
fn main() {
    let _options = Options { bar: "baz".to_string(), ..Default::default() };
}
```

It would be really nice to create non_exhaustive structs using the FRU syntax when a base struct is provided.  I would be okay with a constraint that all fields need to be public.

@Manishearth @zbraniecki

Reply:
A merged RFC is probably not the best place to discuss this: But it could be possible to add `non_exhaustive(pub)` or something with the guarantees required. Unfortunately what you're saying cannot be done with the existing syntax in a backwards compatible way unless FRU syntax is updated to allow for private fields.

Maybe open a thread on https://internals.rust-lang.org/, or discuss it with the lang team on their zulip channel?

Reply:
Done: https://internals.rust-lang.org/t/allow-fru-syntax-on-non-exhaustive-structs/12800